### PR TITLE
[GraphBolt] Add NOT deduplication func to homo graph.

### PR DESCRIPTION
**Pull request by yxy235**
## Description

Add `depulicated` parameter to class `NeighborSampler` , `LayerNeighborSampler` and function `unique_and_compact_node_pairs` to support NOT deduplication when sample >1 hops neighbors.
If `deduplicate` is True, the seeds between hops will be deduplicated. Otherwise, the seeds will be a combination of original seeds and new src_nodes. This function is achieved in `not_deduplication_compact_dst`.

### Example:
**Data**
```
g = dgl.graph(([5, 0, 1, 5, 6, 7, 2, 2, 4], [0, 1, 2, 2, 2, 2, 3, 4, 4]))
graph = gb.from_dglgraph(g, True)

seed_nodes = torch.LongTensor([0, 3, 4])
seed_nodes = torch.LongTensor([5, 2, 2, 4])

itemset = gb.ItemSet(seed_nodes, names="seed_nodes")
item_sampler = gb.ItemSampler(itemset, batch_size=len(seed_nodes))
num_layer = 2
fanouts = [torch.LongTensor([2]) for _ in range(num_layer)]
```
**Deduplicated**
```
datapipe = item_sampler.sample_neighbor(graph, fanouts)
```
**Deduplicated Output**
```
MiniBatch(seed_nodes=tensor([0, 3, 4]),
          sampled_subgraphs=[SampledSubgraphImpl(node_pairs=(tensor([3, 4, 4, 2, 5, 3]), tensor([0, 1, 2, 2, 4, 4])),
                                                original_column_node_ids=tensor([0, 3, 4, 5, 2]),
                                                original_edge_ids=None,
                                                original_row_node_ids=tensor([0, 3, 4, 5, 2, 1]),),
                            SampledSubgraphImpl(node_pairs=(tensor([3, 4, 4, 2]), tensor([0, 1, 2, 2])),
                                                original_column_node_ids=tensor([0, 3, 4]),
                                                original_edge_ids=None,
                                                original_row_node_ids=tensor([0, 3, 4, 5, 2]),)],
          node_pairs=None,
          node_features=None,
          negative_srcs=None,
          negative_dsts=None,
          labels=None,
          input_nodes=tensor([0, 3, 4, 5, 2, 1]),
          edge_features=None,
          compacted_node_pairs=None,
          compacted_negative_srcs=None,
          compacted_negative_dsts=None,
       )
```
**NOT Deduplicated**
```
datapipe = item_sampler.sample_neighbor(graph, fanouts, deduplicated=False)
```
**NOT Deduplicated Output**
```
MiniBatch(seed_nodes=tensor([0, 3, 4]),
          sampled_subgraphs=[SampledSubgraphImpl(node_pairs=(tensor([ 7,  8,  9, 10, 11, 12, 13, 14, 15, 16]), tensor([0, 1, 2, 2, 4, 4, 5, 5, 6, 6])),
                                                original_column_node_ids=tensor([0, 3, 4, 5, 2, 2, 4]),
                                                original_edge_ids=None,
                                                original_row_node_ids=tensor([0, 3, 4, 5, 2, 2, 4, 5, 2, 2, 4, 6, 7, 5, 1, 2, 4]),),
                            SampledSubgraphImpl(node_pairs=(tensor([3, 4, 5, 6]), tensor([0, 1, 2, 2])),
                                                original_column_node_ids=tensor([0, 3, 4]),
                                                original_edge_ids=None,
                                                original_row_node_ids=tensor([0, 3, 4, 5, 2, 2, 4]),)],
          node_pairs=None,
          node_features=None,
          negative_srcs=None,
          negative_dsts=None,
          labels=None,
          input_nodes=tensor([0, 3, 4, 5, 2, 2, 4, 5, 2, 2, 4, 6, 7, 5, 1, 2, 4]),
          edge_features=None,
          compacted_node_pairs=None,
          compacted_negative_srcs=None,
          compacted_negative_dsts=None,
       )
```
## Benchmark
The time includes the same process of generating data.
image
image

The time for `sample_neighbor`, which exclude worst situation.
num_nodes | num_edges | fanouts | deduplication_time (us) | not_deduplication_time (us)
-- | -- | -- | -- | --
6 | 10 | [2, 2] | 123.41 | 75.82
100 | 1000 | [2, 4, 3] | 82.68 | 83.40
10000 | 1000000 | [3, 5, 7, 9, 11] | 148.34 | 146.77
1000000 | 6000000 | [6, 6, 6, 6] | 164.03 | 163.60
1000000 | 10000000 | [13, 14, 15, 16, 17] | 171.28 | 169.04

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 1]
```diff
@@ -267,3 +267,182 @@ def unique_and_compact_node_pairs(
         unique_nodes = list(unique_nodes.values())[0]
 
     return unique_nodes, compacted_node_pairs
+
+
+def duplicated_and_compact_node_pairs(
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/utils/sample_utils.py:272]: the method body and `unique_and_compact_node_pairs` seems have much in common. we'd better avoid code duplication

Comment (frozenbugs) [reply python/dgl/graphbolt/utils/sample_utils.py:272]: deduplicated? not_deduplicated?

Just name it to compact_node_pairs should be fine. Just compact no unique.

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 2]
```diff
@@ -210,14 +210,77 @@ def unique_and_compact_node_pairs(
...
+        Dict[str, Tuple[torch.Tensor, torch.Tensor]],
+    ],
+    seeds: Union[
+        torch.Tensor,
+        Dict[str, torch.Tensor],
+    ] = None,
+    deduplicate=False,
+):
+    """
+    Compact node pairs and return duplicated nodes (per type).
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/utils/sample_utils.py:228]: `return duplicated nodes` is not accurate.

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 3]
```diff
@@ -264,6 +339,71 @@ def unique_and_compact_node_pairs(
     # Return singleton for a homogeneous graph.
     if is_homogeneous:
         compacted_node_pairs = list(compacted_node_pairs.values())[0]
-        unique_nodes = list(unique_nodes.values())[0]
+        cat_seeds = list(cat_seeds.values())[0]
+
+    return cat_seeds, compacted_node_pairs
+
+
+def not_deduplication_compact_dst(
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/utils/sample_utils.py:347]: the name looks non-intuitive to me. `compact_dst_nodes()`?

File: tests/python/pytorch/graphbolt/test_subgraph_sampler.py [hunk 4]
```diff
@@ -192,3 +193,25 @@ def test_SubgraphSampler_Link_Hetero_With_Negative(labor):
...
+    num_layer = 2
+    fanouts = [torch.LongTensor([2]) for _ in range(num_layer)]
+
+    Sampler = gb.LayerNeighborSampler if labor else gb.NeighborSampler
+    datapipe = Sampler(item_sampler, graph, fanouts, deduplicate=False)
+
+    length = [17, 7]
+    for data in datapipe:
+        for step, sampled_subgraph in enumerate(data.sampled_subgraphs):
+            assert len(sampled_subgraph.original_row_node_ids) == length[step]
```
Comment (Rhett-Ying) [inline tests/python/pytorch/graphbolt/test_subgraph_sampler.py:217]: verify internal content data instead of length only is better.

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 5]
```diff
@@ -264,6 +339,71 @@ def unique_and_compact_node_pairs(
...
 
-    return unique_nodes, compacted_node_pairs
+    Parameters
+    ----------
+    dst: torch.Tensor
+        The original dst that need to be compacted.
+    duplicated_dst: torch.Tensor
+        Duplicated_dst record the nodes that generate dst in order. Its index
+        will be used to help generate the compacted dst.
+
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/utils/sample_utils.py:360]: could you add an example here for better readability?

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 6]
```diff
@@ -264,6 +339,71 @@ def unique_and_compact_node_pairs(
...
+    Returns
+    -------
+    torch.Tenor
+        Compacted dst without deduplication.
+    """
+    # Record the number of consecutive identical values.
+    dst_constinuous = [1 for _ in duplicated_dst]
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/utils/sample_utils.py:367]: as this function will be called in critical path, please pay attention on performance.

Comment (Rhett-Ying) [reply python/dgl/graphbolt/utils/sample_utils.py:367]: could you describe the algo and implementation details to help us have a basic and clear understanding.

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 7]
```diff
@@ -228,27 +291,39 @@ def unique_and_compact_node_pairs(
...
+        if deduplicate:
+            seeds = {
+                ntype: torch.unique(nodes) for ntype, nodes in dst_nodes.items()
+            }
+        else:
+            seeds = dst_nodes
 
     ntypes = set(dst_nodes.keys()) | set(src_nodes.keys())
-    unique_nodes = {}
+    cat_seeds = {}
```
Comment (frozenbugs) [inline python/dgl/graphbolt/utils/sample_utils.py:304]: What's cat_seed? cat?

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 8]
```diff
@@ -210,14 +210,77 @@ def unique_and_compact_node_pairs(
...
+def compact_node_pairs(
+    node_pairs: Union[
+        Tuple[torch.Tensor, torch.Tensor],
+        Dict[str, Tuple[torch.Tensor, torch.Tensor]],
+    ],
+    seeds: Union[
+        torch.Tensor,
+        Dict[str, torch.Tensor],
+    ] = None,
+    deduplicate=False,
```
Comment (frozenbugs) [inline python/dgl/graphbolt/utils/sample_utils.py:225]:
```suggestion
    unique=False,
```

File: python/dgl/graphbolt/utils/sample_utils.py (hunk 7 above)
Comment (yxy235) [reply python/dgl/graphbolt/utils/sample_utils.py:304]: The seeds after torch.cat. Maybe can changed into new_seeds?

File: python/dgl/graphbolt/utils/sample_utils.py (hunk 3 above)
Comment (frozenbugs) [inline python/dgl/graphbolt/utils/sample_utils.py:347]: @peizhou001 can you take a look here esp. the efficiency.

Comment (yxy235) [reply python/dgl/graphbolt/utils/sample_utils.py:347]: The time complexity should be O(n). n depends on the length of dst. In order to treat the same elements as different individuals, I haven't thought of a better strategy at the moment.

File: python/dgl/graphbolt/utils/sample_utils.py (hunk 6 above)
Comment (yxy235) [reply python/dgl/graphbolt/utils/sample_utils.py:367]: ## Step 1: Initialize Supporting Structures
    
`seeds_constinuous`: A list tracking consecutive occurrences of elements in seeds.
`dst_compact`: An empty list, later to contain mapped subscripts corresponding to elements in dst.
`compacted_dst_num`: An empty dictionary, which will eventually record the count of appearances of each element in the compacted destination tensor.

**Objective**: Ensure we have structures to store subsequent calculations and mappings.

## Step 2: Establish Initial Mappings

**Action**: Employ a while loop to traverse through `dst` and `seeds` in reverse order.
**Goal**: Populate `dst_compact` and `compacted_dst_num` by establishing a preliminary relationship between corresponding elements in `dst` and `seeds`. Specifically:

- `dst_compact` will contain subscripts pointing to related elements between `dst` and `seeds`.

- `compacted_dst_num` keeps track of how often each subscript appears in `dst_compact`.

**Objective**: Create a rudimentary mapping between dst and the compacted destination tensor, albeit requiring further refinement.

## Step 3: Refine Mappings

**Action**: Traverse dst_compact in reverse once again.
**Goal**: Correct the initial mappings by:

- Checking for sequential duplicates in `dst`.

- Adjusting the mapped values in `dst_compact` to ensure they correctly correspond with the `dst` and `seeds` by calculating match_times and the number of consecutive seeds with same value.

**Objective**: Ensure that each value in `dst` is mapped to a correct, and distinctly corresponding, seed subscript, even when dealing with consecutive identical values.

## Step 4: Finalize and Return Result

**Action**: Convert `dst_compact` into a torch tensor.
**Return**: This torch tensor, which is the refined, compacted destination tensor, `compacted_dst`.

**Objective**: Output a final, compacted tensor that accurately represents dst in a form that’s mapped via subscripts to seeds.

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 9]
```diff
@@ -210,14 +210,78 @@ def unique_and_compact_node_pairs(
...
+        mapped to a contiguous space of IDs ranging from 0 to N.
+
+    Examples
+    --------
+    >>> import dgl.graphbolt as gb
+    >>> N1 = torch.LongTensor([1, 2, 2])
+    >>> N2 = torch.LongTensor([5, 6, 5])
+    >>> node_pairs = {"n1:e1:n2": (N1, N2),
+    ...     "n2:e2:n1": (N2, N1)}
+    >>> duplicated_nodes, compacted_node_pairs = gb.duplicated_and_compact_node_pairs(
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/utils/sample_utils.py:268]: the function name is out-dated

Comment (Rhett-Ying) [reply python/dgl/graphbolt/utils/sample_utils.py:268]: add examples for `unique=True` and `unique=False` respectively.

File: python/dgl/graphbolt/impl/csc_sampling_graph.py [hunk 10]
```diff
@@ -326,6 +326,48 @@ def _convert_to_sampled_subgraph(
             node_pairs=node_pairs, original_edge_ids=original_edge_ids
         )
 
+    def _convert_to_all_sampled_subgraph(
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/csc_sampling_graph.py:329]: What's the difference between this method with _convert_to_sampled_subgraph? This is a pretty poor name, can you come up with a better name, feel free to change _convert_to_sampled_subgraph if you need.

**Comment by frozenbugs** [comment]
Overall LGTM, @peizhou001 can you double check.

File: python/dgl/graphbolt/impl/csc_sampling_graph.py [hunk 11]
```diff
@@ -277,12 +277,13 @@ def in_subgraph(self, nodes: torch.Tensor) -> torch.ScriptObject:
         # TODO: change the result to 'SampledSubgraphImpl'.
         return self._c_csc_graph.in_subgraph(nodes)
 
-    def _convert_to_sampled_subgraph(
+    def _generate_sampled_subgraph_node_pairs_and_eids(
```
Comment (peizhou001) [inline python/dgl/graphbolt/impl/csc_sampling_graph.py:280]: Keep `_convert `in name for consistence with its counterpart  `_convert_to_sampled_subgraph`

File: python/dgl/graphbolt/impl/csc_sampling_graph.py [hunk 12]
```diff
@@ -326,6 +327,48 @@ def _convert_to_sampled_subgraph(
...
+            column = torch.arange(len(column_num)).repeat_interleave(column_num)
+            row = torch.arange(len(C_sampled_subgraph.indices)) + len(
+                column_num
+            )
+            original_edge_ids = C_sampled_subgraph.original_edge_ids
+            has_original_eids = (
+                self.edge_attributes is not None
+                and ORIGINAL_EDGE_ID in self.edge_attributes
+            )
+            if has_original_eids:
```
Comment (peizhou001) [inline python/dgl/graphbolt/impl/csc_sampling_graph.py:350]: also `original_edge_ids  not None`

File: python/dgl/graphbolt/impl/csc_sampling_graph.py [hunk 13]
```diff
@@ -378,6 +422,10 @@ def sample_neighbors(
             corresponding to each neighboring edge of a node. It must be a 1D
             floating-point or boolean tensor, with the number of elements
             equalling the total number of edges.
+        deduplicate: bool
+            Boolean indicating whether seeds between hops will be deduplicated.
+            If True, the same elements in seeds will be deleted to only one.
```
Comment (peizhou001) [inline python/dgl/graphbolt/impl/csc_sampling_graph.py:427]: deleted -> removed

**Review by peizhou001 — APPROVED** [review]
